src/target/riscv/riscv.h

```diff
@@ -129,6 +129,22 @@ struct reg_name_table {
 	char **reg_names;
 };
 
+typedef struct riscv_mem_access_info {
+	target_addr_t target_address;
+
+	union {
+		uint8_t *read;
+		const uint8_t *write;
+	} buffer;
```

Comment:
My preference would be to not use unions because it does not really prevent the code writers from assigning to `buffer.read` but reading back `buffer.write` (or vice versa).

Instead, I'd like to suggest:
- Having separate `uint8* read_buffer` and `const uint8_t* write_buffer`
- Using assertions at places where the data from `riscv_mem_access_info_t` is actually used

This would provide bit more robustness and safety than the current solution.

```suggestion
	uint8_t *read_buffer;
	const uint8_t *write_buffer;
```

```c
// For read functions:
assert(mem_args->is_read);
assert(mem_args->read_buffer);
assert(!mem_args->write_buffer);

// Analogously for write functions.
```

Reply:
> My preference would be to not use unions because it does not really prevent the code writers from assigning to `buffer.read` but reading back `buffer.write` (or vice versa).
> 
> Instead, I'd like to suggest:
> 
> * Having separate `uint8* read_buffer` and `const uint8_t* write_buffer`
> * Using assertions at places where the data from `riscv_mem_access_info_t` is actually used

@JanMatCodasip, I'd like to argue with that.
AFAIU, the approach you suggest will not fix the issue you are talking about.
IMHO, an assertion at the start of the function does not prevent the user from reading the wrong field.
Sure, probably static analysis tools will be more likely to catch the error, but my opinion is: `union` creates less boilerplate (only need to write one field) and therefore an error is less likely.

Anyway, either solution seems fine by me.

Though, if the union is dropped I'd also suggest dropping `is_write` field and creating a function that takes a `riscv_mem_access_info_t` and returns whether it's a write or a read based on buffer pointers (with the necessary assertions).

Reply:
An assertion would however ensure that exactly one of the values is non-NULL. And if the programmer makes a mistake and picks the wrong one, the use of NULL would become apparent very soon.

> if the union is dropped I'd also suggest dropping is_write field and creating a function that takes a riscv_mem_access_info_t and returns whether it's a write or a read based on buffer pointers

Sounds good :)

Reply:
Union removed